// @directus/sdk/dist/handlers/graphql.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GraphQLHandler = void 0;
class GraphQLHandler {
    constructor(transport) {
        this.transport = transport;
    }
    request(base, query, variables) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.transport.post(base, {
                query,
                variables: typeof variables === 'undefined' ? {} : variables,
            });
        });
    }
    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    items(query, variables) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.request('/graphql', query, variables);
        });
    }
    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    system(query, variables) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.request('/graphql/system', query, variables);
        });
    }
}
exports.GraphQLHandler = GraphQLHandler;
