// @directus/sdk/dist/handlers/relations.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RelationsHandler = void 0;
class RelationsHandler {
    constructor(transport) {
        this.transport = transport;
    }
    readOne(collection, id) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.transport.get(`/relations/${collection}/${id}`);
            return response.data;
        });
    }
    readMany(collection) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.transport.get(`/relations/${collection}`);
            return response.data;
        });
    }
    readAll() {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.transport.get(`/relations`);
            return response.data;
        });
    }
    createOne(item) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.transport.post(`/relations`, item)).data;
        });
    }
    updateOne(collection, field, item) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.transport.patch(`/relations/${collection}/${field}`, {
                params: item,
            })).data;
        });
    }
    deleteOne(collection, field) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.transport.delete(`/relations/${collection}/${field}`);
        });
    }
}
exports.RelationsHandler = RelationsHandler;
