// @directus/sdk/dist/handlers/singleton.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SingletonHandler = void 0;
class SingletonHandler {
    constructor(collection, transport) {
        this.collection = collection;
        this.transport = transport;
        this.endpoint = collection.startsWith('directus_') ? `/${collection.substring(9)}` : `/items/${collection}`;
    }
    read(query) {
        return __awaiter(this, void 0, void 0, function* () {
            const item = yield this.transport.get(`${this.endpoint}`, {
                params: query,
            });
            return item.data;
        });
    }
    update(data, _query) {
        return __awaiter(this, void 0, void 0, function* () {
            const item = yield this.transport.patch(`${this.endpoint}`, data, {
                params: _query,
            });
            return item.data;
        });
    }
}
exports.SingletonHandler = SingletonHandler;
