// @directus/sdk/dist/sdk.iife.js
var Directus = (function (exports) {
    'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    class ItemsHandler {
        constructor(collection, transport) {
            this.transport = transport;
            this.endpoint = collection.startsWith('directus_') ? `/${collection.substring(9)}` : `/items/${collection}`;
        }
        readOne(id, query) {
            return __awaiter(this, void 0, void 0, function* () {
                const response = yield this.transport.get(`${this.endpoint}/${encodeURI(id)}`, {
                    params: query,
                });
                return response.data;
            });
        }
        readMany(query) {
            return __awaiter(this, void 0, void 0, function* () {
                const { data, meta } = yield this.transport.get(`${this.endpoint}`, {
                    params: query,
                });
                return {
                    data,
                    meta,
                };
            });
        }
        createOne(item, query) {
            return __awaiter(this, void 0, void 0, function* () {
                return (yield this.transport.post(`${this.endpoint}`, item, {
                    params: query,
                })).data;
            });
        }
        createMany(items, query) {
            return __awaiter(this, void 0, void 0, function* () {
                return yield this.transport.post(`${this.endpoint}`, items, {
                    params: query,
                });
            });
        }
        updateOne(id, item, query) {
            return __awaiter(this, void 0, void 0, function* () {
                return (yield this.transport.patch(`${this.endpoint}/${encodeURI(id)}`, item, {
                    params: query,
                })).data;
            });
        }
        updateMany(ids, data, query) {
            return __awaiter(this, void 0, void 0, function* () {
                return yield this.transport.patch(`${this.endpoint}`, {
                    keys: ids,
                    data,
                }, {
                    params: query,
                });
            });
        }
        updateByQuery(updateQuery, data, query) {
            return __awaiter(this, void 0, void 0, function* () {
                return yield this.transport.patch(`${this.endpoint}`, {
                    query: updateQuery,
                    data,
                }, {
                    params: query,
                });
            });
        }
        deleteOne(id) {
            return __awaiter(this, void 0, void 0, function* () {
                yield this.transport.delete(`${this.endpoint}/${encodeURI(id)}`);
            });
        }
        deleteMany(ids) {
            return __awaiter(this, void 0, void 0, function* () {
                yield this.transport.delete(`${this.endpoint}`, ids);
            });
        }
    }

    class CommentsHandler {
        constructor(transport) {
            this.transport = transport;
        }
        create(comment) {
            return __awaiter(this, void 0, void 0, function* () {
                const response = yield this.transport.post('/activity/comment', comment);
                return response.data;
            });
        }
        update(comment_activity_id, comment) {
            return __awaiter(this, void 0, void 0, function* () {
                const response = yield this.transport.patch(`/activity/comment/${encodeURI(comment_activity_id)}`, {
                    comment,
                });
                return response.data;
            });
        }
        delete(comment_activity_id) {
            return __awaiter(this, void 0, void 0, function* () {
                yield this.transport.delete(`/activity/comment/${encodeURI(comment_activity_id)}`);
            });
        }
    }

    /**
     * Activity handler
     */
    class ActivityHandler extends ItemsHandler {
        constructor(transport) {
            super('directus_activity', transport);
            this._comments = new CommentsHandler(this.transport);
        }
        get comments() {
            return this._comments;
        }
    }

    /**
     * Collections handler
     */
    class CollectionsHandler {
        constructor(transport) {
            this.transport = transport;
        }
        readOne(collection) {
            return __awaiter(this, void 0, void 0, function* () {
                const response = yield this.transport.get(`/collections/${collection}`);
                return response.data;
            });
        }
        readAll() {
            return __awaiter(this, void 0, void 0, function* () {
                const { data, meta } = yield this.transport.get(`/collections`);
                return {
                    data,
                    meta,
                };
            });
        }
        createOne(collection) {
            return __awaiter(this, void 0, void 0, function* () {
                return (yield this.transport.post(`/collections`, collection)).data;
            });
        }
        createMany(collections) {
            return __awaiter(this, void 0, void 0, function* () {
                const { data, meta } = yield this.transport.post(`/collections`, collections);
                return {
                    data,
                    meta,
                };
            });
        }
        updateOne(collection, item, query) {
            return __awaiter(this, void 0, void 0, function* () {
                return (yield this.transport.patch(`/collections/${collection}`, item, {
                    params: query,
                })).data;
            });
        }
        deleteOne(collection) {
            return __awaiter(this, void 0, void 0, function* () {
                yield this.transport.delete(`/collections/${collection}`);
            });
        }
    }

    /**
     * Fields handler
     */
    class FieldsHandler {
        constructor(transport) {
            this.transport = transport;
        }
        readOne(collection, id) {
            return __awaiter(this, void 0, void 0, function* () {
                const response = yield this.transport.get(`/fields/${collection}/${id}`);
                return response.data;
            });
        }
        readMany(collection) {
            return __awaiter(this, void 0, void 0, function* () {
                const response = yield this.transport.get(`/fields/${collection}`);
                return response.data;
            });
        }
        readAll() {
            return __awaiter(this, void 0, void 0, function* () {
                const response = yield this.transport.get(`/fields`);
                return response.data;
            });
        }
        createOne(collection, item) {
            return __awaiter(this, void 0, void 0, function* () {
                return (yield this.transport.post(`/fields/${collection}`, item)).data;
            });
        }
        updateOne(collection, field, item) {
            return __awaiter(this, void 0, void 0, function* () {
                return (yield this.transport.patch(`/fields/${collection}/${field}`, item)).data;
            });
        }
        deleteOne(collection, field) {
            return __awaiter(this, void 0, void 0, function* () {
                yield this.transport.delete(`/fields/${collection}/${field}`);
            });
        }
    }

    /**
     * Files handler
     */
    class FilesHandler extends ItemsHandler {
        constructor(transport) {
            super('directus_files', transport);
        }
    }

    /**
     * Folders handler
     */
    class FoldersHandler extends ItemsHandler {
        constructor(transport) {
            super('directus_folders', transport);
        }
    }

    /**
     * Permissions handler
     */
    class PermissionsHandler extends ItemsHandler {
        constructor(transport) {
            super('directus_permissions', transport);
        }
    }

    /**
     * Presets handler
     */
    class PresetsHandler extends ItemsHandler {
        constructor(transport) {
            super('directus_presets', transport);
        }
    }

    class RelationsHandler {
        constructor(transport) {
            this.transport = transport;
        }
        readOne(collection, id) {
            return __awaiter(this, void 0, void 0, function* () {
                const response = yield this.transport.get(`/relations/${collection}/${id}`);
                return response.data;
            });
        }
        readMany(collection) {
            return __awaiter(this, void 0, void 0, function* () {
                const response = yield this.transport.get(`/relations/${collection}`);
                return response.data;
            });
        }
        readAll() {
            return __awaiter(this, void 0, void 0, function* () {
                const response = yield this.transport.get(`/relations`);
                return response.data;
            });
        }
        createOne(item) {
            return __awaiter(this, void 0, void 0, function* () {
                return (yield this.transport.post(`/relations`, item)).data;
            });
        }
        updateOne(collection, field, item) {
            return __awaiter(this, void 0, void 0, function* () {
                return (yield this.transport.patch(`/relations/${collection}/${field}`, {
                    params: item,
                })).data;
            });
        }
        deleteOne(collection, field) {
            return __awaiter(this, void 0, void 0, function* () {
                yield this.transport.delete(`/relations/${collection}/${field}`);
            });
        }
    }

    /**
     * Revisions handler
     */
    class RevisionsHandler extends ItemsHandler {
        constructor(transport) {
            super('directus_revisions', transport);
        }
    }

    /**
     * Roles handler
     */
    class RolesHandler extends ItemsHandler {
        constructor(transport) {
            super('directus_roles', transport);
        }
    }

    /**
     * Server handler
     */
    class ServerHandler {
        constructor(transport) {
            this.transport = transport;
        }
        ping() {
            return __awaiter(this, void 0, void 0, function* () {
                return (yield this.transport.get('/server/ping')).raw;
            });
        }
        info() {
            return __awaiter(this, void 0, void 0, function* () {
                return (yield this.transport.get('/server/info')).data;
            });
        }
    }

    class SingletonHandler {
        constructor(collection, transport) {
            this.collection = collection;
            this.transport = transport;
            this.endpoint = collection.startsWith('directus_') ? `/${collection.substring(9)}` : `/items/${collection}`;
        }
        read(query) {
            return __awaiter(this, void 0, void 0, function* () {
                const item = yield this.transport.get(`${this.endpoint}`, {
                    params: query,
                });
                return item.data;
            });
        }
        update(data, _query) {
            return __awaiter(this, void 0, void 0, function* () {
                const item = yield this.transport.patch(`${this.endpoint}`, data, {
                    params: _query,
                });
                return item.data;
            });
        }
    }

    class SettingsHandler extends SingletonHandler {
        constructor(transport) {
            super('directus_settings', transport);
        }
    }

    class InvitesHandler {
        constructor(transport) {
            this.transport = transport;
        }
        send(email, role, invite_url) {
            return __awaiter(this, void 0, void 0, function* () {
                yield this.transport.post('/users/invite', {
                    email,
                    role,
                    invite_url,
                });
            });
        }
        accept(token, password) {
            return __awaiter(this, void 0, void 0, function* () {
                yield this.transport.post(`/users/invite/accept`, {
                    token,
                    password,
                });
            });
        }
    }

    class TFAHandler {
        constructor(transport) {
            this.transport = transport;
        }
        generate(password) {
            return __awaiter(this, void 0, void 0, function* () {
                const result = yield this.transport.post('/users/me/tfa/generate', { password });
                return result.data;
            });
        }
        enable(secret, otp) {
            return __awaiter(this, void 0, void 0, function* () {
                yield this.transport.post('/users/me/tfa/enable', { secret, otp });
            });
        }
        disable(otp) {
            return __awaiter(this, void 0, void 0, function* () {
                yield this.transport.post('/users/me/tfa/disable', { otp });
            });
        }
    }

    class MeHandler {
        constructor(transport) {
            this._transport = transport;
        }
        get tfa() {
            return this._tfa || (this._tfa = new TFAHandler(this._transport));
        }
        read(query) {
            return __awaiter(this, void 0, void 0, function* () {
                const response = yield this._transport.get('/users/me', {
                    params: query,
                });
                return response.data;
            });
        }
        update(data, query) {
            return __awaiter(this, void 0, void 0, function* () {
                const response = yield this._transport.patch(`/users/me`, data, {
                    params: query,
                });
                return response.data;
            });
        }
    }

    /**
     * Users handler
     */
    class UsersHandler extends ItemsHandler {
        constructor(transport) {
            super('directus_users', transport);
        }
        get invites() {
            return this._invites || (this._invites = new InvitesHandler(this.transport));
        }
        get me() {
            return this._me || (this._me = new MeHandler(this.transport));
        }
    }

    /**
     * Utils handler
     */
    class UtilsHandler {
        constructor(transport) {
            this.random = {
                string: (length = 32) => __awaiter(this, void 0, void 0, function* () {
                    const result = yield this.transport.get('/utils/random/string', { params: { length } });
                    return result.data;
                }),
            };
            this.hash = {
                generate: (string) => __awaiter(this, void 0, void 0, function* () {
                    const result = yield this.transport.post('/utils/hash/generate', { string });
                    return result.data;
                }),
                verify: (string, hash) => __awaiter(this, void 0, void 0, function* () {
                    const result = yield this.transport.post('/utils/hash/verify', { string, hash });
                    return result.data;
                }),
            };
            this.transport = transport;
        }
        sort(collection, item, to) {
            return __awaiter(this, void 0, void 0, function* () {
                yield this.transport.post(`/utils/sort/${encodeURI(collection)}`, { item, to });
            });
        }
        revert(revision) {
            return __awaiter(this, void 0, void 0, function* () {
                yield this.transport.post(`/utils/revert/${encodeURI(revision)}`);
            });
        }
    }

    exports.Meta = void 0;
    (function (Meta) {
        Meta["TOTAL_COUNT"] = "total_count";
        Meta["FILTER_COUNT"] = "filter_count";
    })(exports.Meta || (exports.Meta = {}));

    class TransportError extends Error {
        constructor(parent, response) {
            var _a, _b;
            if ((_a = response === null || response === void 0 ? void 0 : response.errors) === null || _a === void 0 ? void 0 : _a.length) {
                super((_b = response === null || response === void 0 ? void 0 : response.errors[0]) === null || _b === void 0 ? void 0 : _b.message);
            }
            else {
                super((parent === null || parent === void 0 ? void 0 : parent.message) || 'Unknown transport error');
            }
            this.parent = parent;
            this.response = response;
            this.errors = (response === null || response === void 0 ? void 0 : response.errors) || [];
            if (!Object.values(response || {}).some((value) => value !== undefined)) {
                this.response = undefined;
            }
            Object.setPrototypeOf(this, TransportError.prototype);
        }
    }

    class BaseStorage {
        get auth_token() {
            return this.get('auth_token');
        }
        set auth_token(value) {
            if (value === null) {
                this.delete('auth_token');
            }
            else {
                this.set('auth_token', value);
            }
        }
        get auth_expires() {
            const value = this.get('auth_expires');
            if (value === null) {
                return null;
            }
            return parseInt(value);
        }
        set auth_expires(value) {
            if (value === null) {
                this.delete('auth_expires');
            }
            else {
                this.set('auth_expires', value.toString());
            }
        }
        get auth_refresh_token() {
            return this.get('auth_refresh_token');
        }
        set auth_refresh_token(value) {
            if (value === null) {
                this.delete('auth_refresh_token');
            }
            else {
                this.set('auth_refresh_token', value);
            }
        }
    }

    class MemoryStorage extends BaseStorage {
        constructor(prefix = '') {
            super();
            this.values = {};
            this.prefix = prefix;
        }
        get(key) {
            const k = this.key(key);
            if (k in this.values) {
                return this.values[k];
            }
            return null;
        }
        set(key, value) {
            this.values[this.key(key)] = value;
            return value;
        }
        delete(key) {
            const k = this.key(key);
            const value = this.get(key);
            if (k in this.values) {
                delete this.values[k];
            }
            return value;
        }
        key(name) {
            return `${this.prefix}${name}`;
        }
    }

    class LocalStorage extends BaseStorage {
        constructor(prefix = '') {
            super();
            this.prefix = prefix;
        }
        get(key) {
            const value = localStorage.getItem(this.key(key));
            if (value !== null) {
                return value;
            }
            return null;
        }
        set(key, value) {
            localStorage.setItem(this.key(key), value);
            return value;
        }
        delete(key) {
            const k = this.key(key);
            const value = this.get(k);
            if (value) {
                localStorage.removeItem(k);
            }
            return value;
        }
        key(name) {
            return `${this.prefix}${name}`;
        }
    }

    var axios$2 = {exports: {}};

    var bind$2 = function bind(fn, thisArg) {
      return function wrap() {
        var args = new Array(arguments.length);
        for (var i = 0; i < args.length; i++) {
          args[i] = arguments[i];
        }
        return fn.apply(thisArg, args);
      };
    };

    var bind$1 = bind$2;

    // utils is a library of generic helper functions non-specific to axios

    var toString = Object.prototype.toString;

    /**
     * Determine if a value is an Array
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is an Array, otherwise false
     */
    function isArray(val) {
      return toString.call(val) === '[object Array]';
    }

    /**
     * Determine if a value is undefined
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if the value is undefined, otherwise false
     */
    function isUndefined(val) {
      return typeof val === 'undefined';
    }

    /**
     * Determine if a value is a Buffer
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is a Buffer, otherwise false
     */
    function isBuffer(val) {
      return val !== null && !isUndefined(val) && val.constructor !== null && !isUndefined(val.constructor)
        && typeof val.constructor.isBuffer === 'function' && val.constructor.isBuffer(val);
    }

    /**
     * Determine if a value is an ArrayBuffer
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is an ArrayBuffer, otherwise false
     */
    function isArrayBuffer(val) {
      return toString.call(val) === '[object ArrayBuffer]';
    }

    /**
     * Determine if a value is a FormData
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is an FormData, otherwise false
     */
    function isFormData(val) {
      return (typeof FormData !== 'undefined') && (val instanceof FormData);
    }

    /**
     * Determine if a value is a view on an ArrayBuffer
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is a view on an ArrayBuffer, otherwise false
     */
    function isArrayBufferView(val) {
      var result;
      if ((typeof ArrayBuffer !== 'undefined') && (ArrayBuffer.isView)) {
        result = ArrayBuffer.isView(val);
      } else {
        result = (val) && (val.buffer) && (val.buffer instanceof ArrayBuffer);
      }
      return result;
    }

    /**
     * Determine if a value is a String
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is a String, otherwise false
     */
    function isString(val) {
      return typeof val === 'string';
    }

    /**
     * Determine if a value is a Number
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is a Number, otherwise false
     */
    function isNumber(val) {
      return typeof val === 'number';
    }

    /**
     * Determine if a value is an Object
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is an Object, otherwise false
     */
    function isObject(val) {
      return val !== null && typeof val === 'object';
    }

    /**
     * Determine if a value is a plain Object
     *
     * @param {Object} val The value to test
     * @return {boolean} True if value is a plain Object, otherwise false
     */
    function isPlainObject(val) {
      if (toString.call(val) !== '[object Object]') {
        return false;
      }

      var prototype = Object.getPrototypeOf(val);
      return prototype === null || prototype === Object.prototype;
    }

    /**
     * Determine if a value is a Date
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is a Date, otherwise false
     */
    function isDate(val) {
      return toString.call(val) === '[object Date]';
    }

    /**
     * Determine if a value is a File
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is a File, otherwise false
     */
    function isFile(val) {
      return toString.call(val) === '[object File]';
    }

    /**
     * Determine if a value is a Blob
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is a Blob, otherwise false
     */
    function isBlob(val) {
      return toString.call(val) === '[object Blob]';
    }

    /**
     * Determine if a value is a Function
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is a Function, otherwise false
     */
    function isFunction(val) {
      return toString.call(val) === '[object Function]';
    }

    /**
     * Determine if a value is a Stream
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is a Stream, otherwise false
     */
    function isStream(val) {
      return isObject(val) && isFunction(val.pipe);
    }

    /**
     * Determine if a value is a URLSearchParams object
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is a URLSearchParams object, otherwise false
     */
    function isURLSearchParams(val) {
      return typeof URLSearchParams !== 'undefined' && val instanceof URLSearchParams;
    }

    /**
     * Trim excess whitespace off the beginning and end of a string
     *
     * @param {String} str The String to trim
     * @returns {String} The String freed of excess whitespace
     */
    function trim(str) {
      return str.trim ? str.trim() : str.replace(/^\s+|\s+$/g, '');
    }

    /**
     * Determine if we're running in a standard browser environment
     *
     * This allows axios to run in a web worker, and react-native.
     * Both environments support XMLHttpRequest, but not fully standard globals.
     *
     * web workers:
     *  typeof window -> undefined
     *  typeof document -> undefined
     *
     * react-native:
     *  navigator.product -> 'ReactNative'
     * nativescript
     *  navigator.product -> 'NativeScript' or 'NS'
     */
    function isStandardBrowserEnv() {
      if (typeof navigator !== 'undefined' && (navigator.product === 'ReactNative' ||
                                               navigator.product === 'NativeScript' ||
                                               navigator.product === 'NS')) {
        return false;
      }
      return (
        typeof window !== 'undefined' &&
        typeof document !== 'undefined'
      );
    }

    /**
     * Iterate over an Array or an Object invoking a function for each item.
     *
     * If `obj` is an Array callback will be called passing
     * the value, index, and complete array for each item.
     *
     * If 'obj' is an Object callback will be called passing
     * the value, key, and complete object for each property.
     *
     * @param {Object|Array} obj The object to iterate
     * @param {Function} fn The callback to invoke for each item
     */
    function forEach(obj, fn) {
      // Don't bother if no value provided
      if (obj === null || typeof obj === 'undefined') {
        return;
      }

      // Force an array if not already something iterable
      if (typeof obj !== 'object') {
        /*eslint no-param-reassign:0*/
        obj = [obj];
      }

      if (isArray(obj)) {
        // Iterate over array values
        for (var i = 0, l = obj.length; i < l; i++) {
          fn.call(null, obj[i], i, obj);
        }
      } else {
        // Iterate over object keys
        for (var key in obj) {
          if (Object.prototype.hasOwnProperty.call(obj, key)) {
            fn.call(null, obj[key], key, obj);
          }
        }
      }
    }

    /**
     * Accepts varargs expecting each argument to be an object, then
     * immutably merges the properties of each object and returns result.
     *
     * When multiple objects contain the same key the later object in
     * the arguments list will take precedence.
     *
     * Example:
     *
     * ```js
     * var result = merge({foo: 123}, {foo: 456});
     * console.log(result.foo); // outputs 456
     * ```
     *
     * @param {Object} obj1 Object to merge
     * @returns {Object} Result of all merge properties
     */
    function merge(/* obj1, obj2, obj3, ... */) {
      var result = {};
      function assignValue(val, key) {
        if (isPlainObject(result[key]) && isPlainObject(val)) {
          result[key] = merge(result[key], val);
        } else if (isPlainObject(val)) {
          result[key] = merge({}, val);
        } else if (isArray(val)) {
          result[key] = val.slice();
        } else {
          result[key] = val;
        }
      }

      for (var i = 0, l = arguments.length; i < l; i++) {
        forEach(arguments[i], assignValue);
      }
      return result;
    }

    /**
     * Extends object a by mutably adding to it the properties of object b.
     *
     * @param {Object} a The object to be extended
     * @param {Object} b The object to copy properties from
     * @param {Object} thisArg The object to bind function to
     * @return {Object} The resulting value of object a
     */
    function extend(a, b, thisArg) {
      forEach(b, function assignValue(val, key) {
        if (thisArg && typeof val === 'function') {
          a[key] = bind$1(val, thisArg);
        } else {
          a[key] = val;
        }
      });
      return a;
    }

    /**
     * Remove byte order marker. This catches EF BB BF (the UTF-8 BOM)
     *
     * @param {string} content with BOM
     * @return {string} content value without BOM
     */
    function stripBOM(content) {
      if (content.charCodeAt(0) === 0xFEFF) {
        content = content.slice(1);
      }
      return content;
    }

    var utils$d = {
      isArray: isArray,
      isArrayBuffer: isArrayBuffer,
      isBuffer: isBuffer,
      isFormData: isFormData,
      isArrayBufferView: isArrayBufferView,
      isString: isString,
      isNumber: isNumber,
      isObject: isObject,
      isPlainObject: isPlainObject,
      isUndefined: isUndefined,
      isDate: isDate,
      isFile: isFile,
      isBlob: isBlob,
      isFunction: isFunction,
      isStream: isStream,
      isURLSearchParams: isURLSearchParams,
      isStandardBrowserEnv: isStandardBrowserEnv,
      forEach: forEach,
      merge: merge,
      extend: extend,
      trim: trim,
      stripBOM: stripBOM
    };

    var utils$c = utils$d;

    function encode(val) {
      return encodeURIComponent(val).
        replace(/%3A/gi, ':').
        replace(/%24/g, '$').
        replace(/%2C/gi, ',').
        replace(/%20/g, '+').
        replace(/%5B/gi, '[').
        replace(/%5D/gi, ']');
    }

    /**
     * Build a URL by appending params to the end
     *
     * @param {string} url The base of the url (e.g., http://www.google.com)
     * @param {object} [params] The params to be appended
     * @returns {string} The formatted url
     */
    var buildURL$2 = function buildURL(url, params, paramsSerializer) {
      /*eslint no-param-reassign:0*/
      if (!params) {
        return url;
      }

      var serializedParams;
      if (paramsSerializer) {
        serializedParams = paramsSerializer(params);
      } else if (utils$c.isURLSearchParams(params)) {
        serializedParams = params.toString();
      } else {
        var parts = [];

        utils$c.forEach(params, function serialize(val, key) {
          if (val === null || typeof val === 'undefined') {
            return;
          }

          if (utils$c.isArray(val)) {
            key = key + '[]';
          } else {
            val = [val];
          }

          utils$c.forEach(val, function parseValue(v) {
            if (utils$c.isDate(v)) {
              v = v.toISOString();
            } else if (utils$c.isObject(v)) {
              v = JSON.stringify(v);
            }
            parts.push(encode(key) + '=' + encode(v));
          });
        });

        serializedParams = parts.join('&');
      }

      if (serializedParams) {
        var hashmarkIndex = url.indexOf('#');
        if (hashmarkIndex !== -1) {
          url = url.slice(0, hashmarkIndex);
        }

        url += (url.indexOf('?') === -1 ? '?' : '&') + serializedParams;
      }

      return url;
    };

    var utils$b = utils$d;

    function InterceptorManager$1() {
      this.handlers = [];
    }

    /**
     * Add a new interceptor to the stack
     *
     * @param {Function} fulfilled The function to handle `then` for a `Promise`
     * @param {Function} rejected The function to handle `reject` for a `Promise`
     *
     * @return {Number} An ID used to remove interceptor later
     */
    InterceptorManager$1.prototype.use = function use(fulfilled, rejected, options) {
      this.handlers.push({
        fulfilled: fulfilled,
        rejected: rejected,
        synchronous: options ? options.synchronous : false,
        runWhen: options ? options.runWhen : null
      });
      return this.handlers.length - 1;
    };

    /**
     * Remove an interceptor from the stack
     *
     * @param {Number} id The ID that was returned by `use`
     */
    InterceptorManager$1.prototype.eject = function eject(id) {
      if (this.handlers[id]) {
        this.handlers[id] = null;
      }
    };

    /**
     * Iterate over all the registered interceptors
     *
     * This method is particularly useful for skipping over any
     * interceptors that may have become `null` calling `eject`.
     *
     * @param {Function} fn The function to call for each interceptor
     */
    InterceptorManager$1.prototype.forEach = function forEach(fn) {
      utils$b.forEach(this.handlers, function forEachHandler(h) {
        if (h !== null) {
          fn(h);
        }
      });
    };

    var InterceptorManager_1 = InterceptorManager$1;

    var utils$a = utils$d;

    var normalizeHeaderName$1 = function normalizeHeaderName(headers, normalizedName) {
      utils$a.forEach(headers, function processHeader(value, name) {
        if (name !== normalizedName && name.toUpperCase() === normalizedName.toUpperCase()) {
          headers[normalizedName] = value;
          delete headers[name];
        }
      });
    };

    /**
     * Update an Error with the specified config, error code, and response.
     *
     * @param {Error} error The error to update.
     * @param {Object} config The config.
     * @param {string} [code] The error code (for example, 'ECONNABORTED').
     * @param {Object} [request] The request.
     * @param {Object} [response] The response.
     * @returns {Error} The error.
     */
    var enhanceError$2 = function enhanceError(error, config, code, request, response) {
      error.config = config;
      if (code) {
        error.code = code;
      }

      error.request = request;
      error.response = response;
      error.isAxiosError = true;

      error.toJSON = function toJSON() {
        return {
          // Standard
          message: this.message,
          name: this.name,
          // Microsoft
          description: this.description,
          number: this.number,
          // Mozilla
          fileName: this.fileName,
          lineNumber: this.lineNumber,
          columnNumber: this.columnNumber,
          stack: this.stack,
          // Axios
          config: this.config,
          code: this.code,
          status: this.response && this.response.status ? this.response.status : null
        };
      };
      return error;
    };

    var enhanceError$1 = enhanceError$2;

    /**
     * Create an Error with the specified message, config, error code, request and response.
     *
     * @param {string} message The error message.
     * @param {Object} config The config.
     * @param {string} [code] The error code (for example, 'ECONNABORTED').
     * @param {Object} [request] The request.
     * @param {Object} [response] The response.
     * @returns {Error} The created error.
     */
    var createError$2 = function createError(message, config, code, request, response) {
      var error = new Error(message);
      return enhanceError$1(error, config, code, request, response);
    };

    var createError$1 = createError$2;

    /**
     * Resolve or reject a Promise based on response status.
     *
     * @param {Function} resolve A function that resolves the promise.
     * @param {Function} reject A function that rejects the promise.
     * @param {object} response The response.
     */
    var settle$1 = function settle(resolve, reject, response) {
      var validateStatus = response.config.validateStatus;
      if (!response.status || !validateStatus || validateStatus(response.status)) {
        resolve(response);
      } else {
        reject(createError$1(
          'Request failed with status code ' + response.status,
          response.config,
          null,
          response.request,
          response
        ));
      }
    };

    var utils$9 = utils$d;

    var cookies$1 = (
      utils$9.isStandardBrowserEnv() ?

      // Standard browser envs support document.cookie
        (function standardBrowserEnv() {
          return {
            write: function write(name, value, expires, path, domain, secure) {
              var cookie = [];
              cookie.push(name + '=' + encodeURIComponent(value));

              if (utils$9.isNumber(expires)) {
                cookie.push('expires=' + new Date(expires).toGMTString());
              }

              if (utils$9.isString(path)) {
                cookie.push('path=' + path);
              }

              if (utils$9.isString(domain)) {
                cookie.push('domain=' + domain);
              }

              if (secure === true) {
                cookie.push('secure');
              }

              document.cookie = cookie.join('; ');
            },

            read: function read(name) {
              var match = document.cookie.match(new RegExp('(^|;\\s*)(' + name + ')=([^;]*)'));
              return (match ? decodeURIComponent(match[3]) : null);
            },

            remove: function remove(name) {
              this.write(name, '', Date.now() - 86400000);
            }
          };
        })() :

      // Non standard browser env (web workers, react-native) lack needed support.
        (function nonStandardBrowserEnv() {
          return {
            write: function write() {},
            read: function read() { return null; },
            remove: function remove() {}
          };
        })()
    );

    /**
     * Determines whether the specified URL is absolute
     *
     * @param {string} url The URL to test
     * @returns {boolean} True if the specified URL is absolute, otherwise false
     */
    var isAbsoluteURL$1 = function isAbsoluteURL(url) {
      // A URL is considered absolute if it begins with "<scheme>://" or "//" (protocol-relative URL).
      // RFC 3986 defines scheme name as a sequence of characters beginning with a letter and followed
      // by any combination of letters, digits, plus, period, or hyphen.
      return /^([a-z][a-z\d\+\-\.]*:)?\/\//i.test(url);
    };

    /**
     * Creates a new URL by combining the specified URLs
     *
     * @param {string} baseURL The base URL
     * @param {string} relativeURL The relative URL
     * @returns {string} The combined URL
     */
    var combineURLs$1 = function combineURLs(baseURL, relativeURL) {
      return relativeURL
        ? baseURL.replace(/\/+$/, '') + '/' + relativeURL.replace(/^\/+/, '')
        : baseURL;
    };

    var isAbsoluteURL = isAbsoluteURL$1;
    var combineURLs = combineURLs$1;

    /**
     * Creates a new URL by combining the baseURL with the requestedURL,
     * only when the requestedURL is not already an absolute URL.
     * If the requestURL is absolute, this function returns the requestedURL untouched.
     *
     * @param {string} baseURL The base URL
     * @param {string} requestedURL Absolute or relative URL to combine
     * @returns {string} The combined full path
     */
    var buildFullPath$1 = function buildFullPath(baseURL, requestedURL) {
      if (baseURL && !isAbsoluteURL(requestedURL)) {
        return combineURLs(baseURL, requestedURL);
      }
      return requestedURL;
    };

    var utils$8 = utils$d;

    // Headers whose duplicates are ignored by node
    // c.f. https://nodejs.org/api/http.html#http_message_headers
    var ignoreDuplicateOf = [
      'age', 'authorization', 'content-length', 'content-type', 'etag',
      'expires', 'from', 'host', 'if-modified-since', 'if-unmodified-since',
      'last-modified', 'location', 'max-forwards', 'proxy-authorization',
      'referer', 'retry-after', 'user-agent'
    ];

    /**
     * Parse headers into an object
     *
     * ```
     * Date: Wed, 27 Aug 2014 08:58:49 GMT
     * Content-Type: application/json
     * Connection: keep-alive
     * Transfer-Encoding: chunked
     * ```
     *
     * @param {String} headers Headers needing to be parsed
     * @returns {Object} Headers parsed into an object
     */
    var parseHeaders$1 = function parseHeaders(headers) {
      var parsed = {};
      var key;
      var val;
      var i;

      if (!headers) { return parsed; }

      utils$8.forEach(headers.split('\n'), function parser(line) {
        i = line.indexOf(':');
        key = utils$8.trim(line.substr(0, i)).toLowerCase();
        val = utils$8.trim(line.substr(i + 1));

        if (key) {
          if (parsed[key] && ignoreDuplicateOf.indexOf(key) >= 0) {
            return;
          }
          if (key === 'set-cookie') {
            parsed[key] = (parsed[key] ? parsed[key] : []).concat([val]);
          } else {
            parsed[key] = parsed[key] ? parsed[key] + ', ' + val : val;
          }
        }
      });

      return parsed;
    };

    var utils$7 = utils$d;

    var isURLSameOrigin$1 = (
      utils$7.isStandardBrowserEnv() ?

      // Standard browser envs have full support of the APIs needed to test
      // whether the request URL is of the same origin as current location.
        (function standardBrowserEnv() {
          var msie = /(msie|trident)/i.test(navigator.userAgent);
          var urlParsingNode = document.createElement('a');
          var originURL;

          /**
        * Parse a URL to discover it's components
        *
        * @param {String} url The URL to be parsed
        * @returns {Object}
        */
          function resolveURL(url) {
            var href = url;

            if (msie) {
            // IE needs attribute set twice to normalize properties
              urlParsingNode.setAttribute('href', href);
              href = urlParsingNode.href;
            }

            urlParsingNode.setAttribute('href', href);

            // urlParsingNode provides the UrlUtils interface - http://url.spec.whatwg.org/#urlutils
            return {
              href: urlParsingNode.href,
              protocol: urlParsingNode.protocol ? urlParsingNode.protocol.replace(/:$/, '') : '',
              host: urlParsingNode.host,
              search: urlParsingNode.search ? urlParsingNode.search.replace(/^\?/, '') : '',
              hash: urlParsingNode.hash ? urlParsingNode.hash.replace(/^#/, '') : '',
              hostname: urlParsingNode.hostname,
              port: urlParsingNode.port,
              pathname: (urlParsingNode.pathname.charAt(0) === '/') ?
                urlParsingNode.pathname :
                '/' + urlParsingNode.pathname
            };
          }

          originURL = resolveURL(window.location.href);

          /**
        * Determine if a URL shares the same origin as the current location
        *
        * @param {String} requestURL The URL to test
        * @returns {boolean} True if URL shares the same origin, otherwise false
        */
          return function isURLSameOrigin(requestURL) {
            var parsed = (utils$7.isString(requestURL)) ? resolveURL(requestURL) : requestURL;
            return (parsed.protocol === originURL.protocol &&
                parsed.host === originURL.host);
          };
        })() :

      // Non standard browser envs (web workers, react-native) lack needed support.
        (function nonStandardBrowserEnv() {
          return function isURLSameOrigin() {
            return true;
          };
        })()
    );

    /**
     * A `Cancel` is an object that is thrown when an operation is canceled.
     *
     * @class
     * @param {string=} message The message.
     */
    function Cancel$3(message) {
      this.message = message;
    }

    Cancel$3.prototype.toString = function toString() {
      return 'Cancel' + (this.message ? ': ' + this.message : '');
    };

    Cancel$3.prototype.__CANCEL__ = true;

    var Cancel_1 = Cancel$3;

    var utils$6 = utils$d;
    var settle = settle$1;
    var cookies = cookies$1;
    var buildURL$1 = buildURL$2;
    var buildFullPath = buildFullPath$1;
    var parseHeaders = parseHeaders$1;
    var isURLSameOrigin = isURLSameOrigin$1;
    var createError = createError$2;
    var defaults$4 = defaults_1;
    var Cancel$2 = Cancel_1;

    var xhr = function xhrAdapter(config) {
      return new Promise(function dispatchXhrRequest(resolve, reject) {
        var requestData = config.data;
        var requestHeaders = config.headers;
        var responseType = config.responseType;
        var onCanceled;
        function done() {
          if (config.cancelToken) {
            config.cancelToken.unsubscribe(onCanceled);
          }

          if (config.signal) {
            config.signal.removeEventListener('abort', onCanceled);
          }
        }

        if (utils$6.isFormData(requestData)) {
          delete requestHeaders['Content-Type']; // Let the browser set it
        }

        var request = new XMLHttpRequest();

        // HTTP basic authentication
        if (config.auth) {
          var username = config.auth.username || '';
          var password = config.auth.password ? unescape(encodeURIComponent(config.auth.password)) : '';
          requestHeaders.Authorization = 'Basic ' + btoa(username + ':' + password);
        }

        var fullPath = buildFullPath(config.baseURL, config.url);
        request.open(config.method.toUpperCase(), buildURL$1(fullPath, config.params, config.paramsSerializer), true);

        // Set the request timeout in MS
        request.timeout = config.timeout;

        function onloadend() {
          if (!request) {
            return;
          }
          // Prepare the response
          var responseHeaders = 'getAllResponseHeaders' in request ? parseHeaders(request.getAllResponseHeaders()) : null;
          var responseData = !responseType || responseType === 'text' ||  responseType === 'json' ?
            request.responseText : request.response;
          var response = {
            data: responseData,
            status: request.status,
            statusText: request.statusText,
            headers: responseHeaders,
            config: config,
            request: request
          };

          settle(function _resolve(value) {
            resolve(value);
            done();
          }, function _reject(err) {
            reject(err);
            done();
          }, response);

          // Clean up request
          request = null;
        }

        if ('onloadend' in request) {
          // Use onloadend if available
          request.onloadend = onloadend;
        } else {
          // Listen for ready state to emulate onloadend
          request.onreadystatechange = function handleLoad() {
            if (!request || request.readyState !== 4) {
              return;
            }

            // The request errored out and we didn't get a response, this will be
            // handled by onerror instead
            // With one exception: request that using file: protocol, most browsers
            // will return status as 0 even though it's a successful request
            if (request.status === 0 && !(request.responseURL && request.responseURL.indexOf('file:') === 0)) {
              return;
            }
            // readystate handler is calling before onerror or ontimeout handlers,
            // so we should call onloadend on the next 'tick'
            setTimeout(onloadend);
          };
        }

        // Handle browser request cancellation (as opposed to a manual cancellation)
        request.onabort = function handleAbort() {
          if (!request) {
            return;
          }

          reject(createError('Request aborted', config, 'ECONNABORTED', request));

          // Clean up request
          request = null;
        };

        // Handle low level network errors
        request.onerror = function handleError() {
          // Real errors are hidden from us by the browser
          // onerror should only fire if it's a network error
          reject(createError('Network Error', config, null, request));

          // Clean up request
          request = null;
        };

        // Handle timeout
        request.ontimeout = function handleTimeout() {
          var timeoutErrorMessage = 'timeout of ' + config.timeout + 'ms exceeded';
          var transitional = config.transitional || defaults$4.transitional;
          if (config.timeoutErrorMessage) {
            timeoutErrorMessage = config.timeoutErrorMessage;
          }
          reject(createError(
            timeoutErrorMessage,
            config,
            transitional.clarifyTimeoutError ? 'ETIMEDOUT' : 'ECONNABORTED',
            request));

          // Clean up request
          request = null;
        };

        // Add xsrf header
        // This is only done if running in a standard browser environment.
        // Specifically not if we're in a web worker, or react-native.
        if (utils$6.isStandardBrowserEnv()) {
          // Add xsrf header
          var xsrfValue = (config.withCredentials || isURLSameOrigin(fullPath)) && config.xsrfCookieName ?
            cookies.read(config.xsrfCookieName) :
            undefined;

          if (xsrfValue) {
            requestHeaders[config.xsrfHeaderName] = xsrfValue;
          }
        }

        // Add headers to the request
        if ('setRequestHeader' in request) {
          utils$6.forEach(requestHeaders, function setRequestHeader(val, key) {
            if (typeof requestData === 'undefined' && key.toLowerCase() === 'content-type') {
              // Remove Content-Type if data is undefined
              delete requestHeaders[key];
            } else {
              // Otherwise add header to the request
              request.setRequestHeader(key, val);
            }
          });
        }

        // Add withCredentials to request if needed
        if (!utils$6.isUndefined(config.withCredentials)) {
          request.withCredentials = !!config.withCredentials;
        }

        // Add responseType to request if needed
        if (responseType && responseType !== 'json') {
          request.responseType = config.responseType;
        }

        // Handle progress if needed
        if (typeof config.onDownloadProgress === 'function') {
          request.addEventListener('progress', config.onDownloadProgress);
        }

        // Not all browsers support upload events
        if (typeof config.onUploadProgress === 'function' && request.upload) {
          request.upload.addEventListener('progress', config.onUploadProgress);
        }

        if (config.cancelToken || config.signal) {
          // Handle cancellation
          // eslint-disable-next-line func-names
          onCanceled = function(cancel) {
            if (!request) {
              return;
            }
            reject(!cancel || (cancel && cancel.type) ? new Cancel$2('canceled') : cancel);
            request.abort();
            request = null;
          };

          config.cancelToken && config.cancelToken.subscribe(onCanceled);
          if (config.signal) {
            config.signal.aborted ? onCanceled() : config.signal.addEventListener('abort', onCanceled);
          }
        }

        if (!requestData) {
          requestData = null;
        }

        // Send the request
        request.send(requestData);
      });
    };

    var utils$5 = utils$d;
    var normalizeHeaderName = normalizeHeaderName$1;
    var enhanceError = enhanceError$2;

    var DEFAULT_CONTENT_TYPE = {
      'Content-Type': 'application/x-www-form-urlencoded'
    };

    function setContentTypeIfUnset(headers, value) {
      if (!utils$5.isUndefined(headers) && utils$5.isUndefined(headers['Content-Type'])) {
        headers['Content-Type'] = value;
      }
    }

    function getDefaultAdapter() {
      var adapter;
      if (typeof XMLHttpRequest !== 'undefined') {
        // For browsers use XHR adapter
        adapter = xhr;
      } else if (typeof process !== 'undefined' && Object.prototype.toString.call(process) === '[object process]') {
        // For node use HTTP adapter
        adapter = xhr;
      }
      return adapter;
    }

    function stringifySafely(rawValue, parser, encoder) {
      if (utils$5.isString(rawValue)) {
        try {
          (parser || JSON.parse)(rawValue);
          return utils$5.trim(rawValue);
        } catch (e) {
          if (e.name !== 'SyntaxError') {
            throw e;
          }
        }
      }

      return (encoder || JSON.stringify)(rawValue);
    }

    var defaults$3 = {

      transitional: {
        silentJSONParsing: true,
        forcedJSONParsing: true,
        clarifyTimeoutError: false
      },

      adapter: getDefaultAdapter(),

      transformRequest: [function transformRequest(data, headers) {
        normalizeHeaderName(headers, 'Accept');
        normalizeHeaderName(headers, 'Content-Type');

        if (utils$5.isFormData(data) ||
          utils$5.isArrayBuffer(data) ||
          utils$5.isBuffer(data) ||
          utils$5.isStream(data) ||
          utils$5.isFile(data) ||
          utils$5.isBlob(data)
        ) {
          return data;
        }
        if (utils$5.isArrayBufferView(data)) {
          return data.buffer;
        }
        if (utils$5.isURLSearchParams(data)) {
          setContentTypeIfUnset(headers, 'application/x-www-form-urlencoded;charset=utf-8');
          return data.toString();
        }
        if (utils$5.isObject(data) || (headers && headers['Content-Type'] === 'application/json')) {
          setContentTypeIfUnset(headers, 'application/json');
          return stringifySafely(data);
        }
        return data;
      }],

      transformResponse: [function transformResponse(data) {
        var transitional = this.transitional || defaults$3.transitional;
        var silentJSONParsing = transitional && transitional.silentJSONParsing;
        var forcedJSONParsing = transitional && transitional.forcedJSONParsing;
        var strictJSONParsing = !silentJSONParsing && this.responseType === 'json';

        if (strictJSONParsing || (forcedJSONParsing && utils$5.isString(data) && data.length)) {
          try {
            return JSON.parse(data);
          } catch (e) {
            if (strictJSONParsing) {
              if (e.name === 'SyntaxError') {
                throw enhanceError(e, this, 'E_JSON_PARSE');
              }
              throw e;
            }
          }
        }

        return data;
      }],

      /**
       * A timeout in milliseconds to abort a request. If set to 0 (default) a
       * timeout is not created.
       */
      timeout: 0,

      xsrfCookieName: 'XSRF-TOKEN',
      xsrfHeaderName: 'X-XSRF-TOKEN',

      maxContentLength: -1,
      maxBodyLength: -1,

      validateStatus: function validateStatus(status) {
        return status >= 200 && status < 300;
      },

      headers: {
        common: {
          'Accept': 'application/json, text/plain, */*'
        }
      }
    };

    utils$5.forEach(['delete', 'get', 'head'], function forEachMethodNoData(method) {
      defaults$3.headers[method] = {};
    });

    utils$5.forEach(['post', 'put', 'patch'], function forEachMethodWithData(method) {
      defaults$3.headers[method] = utils$5.merge(DEFAULT_CONTENT_TYPE);
    });

    var defaults_1 = defaults$3;

    var utils$4 = utils$d;
    var defaults$2 = defaults_1;

    /**
     * Transform the data for a request or a response
     *
     * @param {Object|String} data The data to be transformed
     * @param {Array} headers The headers for the request or response
     * @param {Array|Function} fns A single function or Array of functions
     * @returns {*} The resulting transformed data
     */
    var transformData$1 = function transformData(data, headers, fns) {
      var context = this || defaults$2;
      /*eslint no-param-reassign:0*/
      utils$4.forEach(fns, function transform(fn) {
        data = fn.call(context, data, headers);
      });

      return data;
    };

    var isCancel$1 = function isCancel(value) {
      return !!(value && value.__CANCEL__);
    };

    var utils$3 = utils$d;
    var transformData = transformData$1;
    var isCancel = isCancel$1;
    var defaults$1 = defaults_1;
    var Cancel$1 = Cancel_1;

    /**
     * Throws a `Cancel` if cancellation has been requested.
     */
    function throwIfCancellationRequested(config) {
      if (config.cancelToken) {
        config.cancelToken.throwIfRequested();
      }

      if (config.signal && config.signal.aborted) {
        throw new Cancel$1('canceled');
      }
    }

    /**
     * Dispatch a request to the server using the configured adapter.
     *
     * @param {object} config The config that is to be used for the request
     * @returns {Promise} The Promise to be fulfilled
     */
    var dispatchRequest$1 = function dispatchRequest(config) {
      throwIfCancellationRequested(config);

      // Ensure headers exist
      config.headers = config.headers || {};

      // Transform request data
      config.data = transformData.call(
        config,
        config.data,
        config.headers,
        config.transformRequest
      );

      // Flatten headers
      config.headers = utils$3.merge(
        config.headers.common || {},
        config.headers[config.method] || {},
        config.headers
      );

      utils$3.forEach(
        ['delete', 'get', 'head', 'post', 'put', 'patch', 'common'],
        function cleanHeaderConfig(method) {
          delete config.headers[method];
        }
      );

      var adapter = config.adapter || defaults$1.adapter;

      return adapter(config).then(function onAdapterResolution(response) {
        throwIfCancellationRequested(config);

        // Transform response data
        response.data = transformData.call(
          config,
          response.data,
          response.headers,
          config.transformResponse
        );

        return response;
      }, function onAdapterRejection(reason) {
        if (!isCancel(reason)) {
          throwIfCancellationRequested(config);

          // Transform response data
          if (reason && reason.response) {
            reason.response.data = transformData.call(
              config,
              reason.response.data,
              reason.response.headers,
              config.transformResponse
            );
          }
        }

        return Promise.reject(reason);
      });
    };

    var utils$2 = utils$d;

    /**
     * Config-specific merge-function which creates a new config-object
     * by merging two configuration objects together.
     *
     * @param {Object} config1
     * @param {Object} config2
     * @returns {Object} New object resulting from merging config2 to config1
     */
    var mergeConfig$2 = function mergeConfig(config1, config2) {
      // eslint-disable-next-line no-param-reassign
      config2 = config2 || {};
      var config = {};

      function getMergedValue(target, source) {
        if (utils$2.isPlainObject(target) && utils$2.isPlainObject(source)) {
          return utils$2.merge(target, source);
        } else if (utils$2.isPlainObject(source)) {
          return utils$2.merge({}, source);
        } else if (utils$2.isArray(source)) {
          return source.slice();
        }
        return source;
      }

      // eslint-disable-next-line consistent-return
      function mergeDeepProperties(prop) {
        if (!utils$2.isUndefined(config2[prop])) {
          return getMergedValue(config1[prop], config2[prop]);
        } else if (!utils$2.isUndefined(config1[prop])) {
          return getMergedValue(undefined, config1[prop]);
        }
      }

      // eslint-disable-next-line consistent-return
      function valueFromConfig2(prop) {
        if (!utils$2.isUndefined(config2[prop])) {
          return getMergedValue(undefined, config2[prop]);
        }
      }

      // eslint-disable-next-line consistent-return
      function defaultToConfig2(prop) {
        if (!utils$2.isUndefined(config2[prop])) {
          return getMergedValue(undefined, config2[prop]);
        } else if (!utils$2.isUndefined(config1[prop])) {
          return getMergedValue(undefined, config1[prop]);
        }
      }

      // eslint-disable-next-line consistent-return
      function mergeDirectKeys(prop) {
        if (prop in config2) {
          return getMergedValue(config1[prop], config2[prop]);
        } else if (prop in config1) {
          return getMergedValue(undefined, config1[prop]);
        }
      }

      var mergeMap = {
        'url': valueFromConfig2,
        'method': valueFromConfig2,
        'data': valueFromConfig2,
        'baseURL': defaultToConfig2,
        'transformRequest': defaultToConfig2,
        'transformResponse': defaultToConfig2,
        'paramsSerializer': defaultToConfig2,
        'timeout': defaultToConfig2,
        'timeoutMessage': defaultToConfig2,
        'withCredentials': defaultToConfig2,
        'adapter': defaultToConfig2,
        'responseType': defaultToConfig2,
        'xsrfCookieName': defaultToConfig2,
        'xsrfHeaderName': defaultToConfig2,
        'onUploadProgress': defaultToConfig2,
        'onDownloadProgress': defaultToConfig2,
        'decompress': defaultToConfig2,
        'maxContentLength': defaultToConfig2,
        'maxBodyLength': defaultToConfig2,
        'transport': defaultToConfig2,
        'httpAgent': defaultToConfig2,
        'httpsAgent': defaultToConfig2,
        'cancelToken': defaultToConfig2,
        'socketPath': defaultToConfig2,
        'responseEncoding': defaultToConfig2,
        'validateStatus': mergeDirectKeys
      };

      utils$2.forEach(Object.keys(config1).concat(Object.keys(config2)), function computeConfigValue(prop) {
        var merge = mergeMap[prop] || mergeDeepProperties;
        var configValue = merge(prop);
        (utils$2.isUndefined(configValue) && merge !== mergeDirectKeys) || (config[prop] = configValue);
      });

      return config;
    };

    var data = {
      "version": "0.22.0"
    };

    var VERSION = data.version;

    var validators$1 = {};

    // eslint-disable-next-line func-names
    ['object', 'boolean', 'number', 'function', 'string', 'symbol'].forEach(function(type, i) {
      validators$1[type] = function validator(thing) {
        return typeof thing === type || 'a' + (i < 1 ? 'n ' : ' ') + type;
      };
    });

    var deprecatedWarnings = {};

    /**
     * Transitional option validator
     * @param {function|boolean?} validator - set to false if the transitional option has been removed
     * @param {string?} version - deprecated version / removed since version
     * @param {string?} message - some message with additional info
     * @returns {function}
     */
    validators$1.transitional = function transitional(validator, version, message) {
      function formatMessage(opt, desc) {
        return '[Axios v' + VERSION + '] Transitional option \'' + opt + '\'' + desc + (message ? '. ' + message : '');
      }

      // eslint-disable-next-line func-names
      return function(value, opt, opts) {
        if (validator === false) {
          throw new Error(formatMessage(opt, ' has been removed' + (version ? ' in ' + version : '')));
        }

        if (version && !deprecatedWarnings[opt]) {
          deprecatedWarnings[opt] = true;
          // eslint-disable-next-line no-console
          console.warn(
            formatMessage(
              opt,
              ' has been deprecated since v' + version + ' and will be removed in the near future'
            )
          );
        }

        return validator ? validator(value, opt, opts) : true;
      };
    };

    /**
     * Assert object's properties type
     * @param {object} options
     * @param {object} schema
     * @param {boolean?} allowUnknown
     */

    function assertOptions(options, schema, allowUnknown) {
      if (typeof options !== 'object') {
        throw new TypeError('options must be an object');
      }
      var keys = Object.keys(options);
      var i = keys.length;
      while (i-- > 0) {
        var opt = keys[i];
        var validator = schema[opt];
        if (validator) {
          var value = options[opt];
          var result = value === undefined || validator(value, opt, options);
          if (result !== true) {
            throw new TypeError('option ' + opt + ' must be ' + result);
          }
          continue;
        }
        if (allowUnknown !== true) {
          throw Error('Unknown option ' + opt);
        }
      }
    }

    var validator$1 = {
      assertOptions: assertOptions,
      validators: validators$1
    };

    var utils$1 = utils$d;
    var buildURL = buildURL$2;
    var InterceptorManager = InterceptorManager_1;
    var dispatchRequest = dispatchRequest$1;
    var mergeConfig$1 = mergeConfig$2;
    var validator = validator$1;

    var validators = validator.validators;
    /**
     * Create a new instance of Axios
     *
     * @param {Object} instanceConfig The default config for the instance
     */
    function Axios$1(instanceConfig) {
      this.defaults = instanceConfig;
      this.interceptors = {
        request: new InterceptorManager(),
        response: new InterceptorManager()
      };
    }

    /**
     * Dispatch a request
     *
     * @param {Object} config The config specific for this request (merged with this.defaults)
     */
    Axios$1.prototype.request = function request(config) {
      /*eslint no-param-reassign:0*/
      // Allow for axios('example/url'[, config]) a la fetch API
      if (typeof config === 'string') {
        config = arguments[1] || {};
        config.url = arguments[0];
      } else {
        config = config || {};
      }

      config = mergeConfig$1(this.defaults, config);

      // Set config.method
      if (config.method) {
        config.method = config.method.toLowerCase();
      } else if (this.defaults.method) {
        config.method = this.defaults.method.toLowerCase();
      } else {
        config.method = 'get';
      }

      var transitional = config.transitional;

      if (transitional !== undefined) {
        validator.assertOptions(transitional, {
          silentJSONParsing: validators.transitional(validators.boolean),
          forcedJSONParsing: validators.transitional(validators.boolean),
          clarifyTimeoutError: validators.transitional(validators.boolean)
        }, false);
      }

      // filter out skipped interceptors
      var requestInterceptorChain = [];
      var synchronousRequestInterceptors = true;
      this.interceptors.request.forEach(function unshiftRequestInterceptors(interceptor) {
        if (typeof interceptor.runWhen === 'function' && interceptor.runWhen(config) === false) {
          return;
        }

        synchronousRequestInterceptors = synchronousRequestInterceptors && interceptor.synchronous;

        requestInterceptorChain.unshift(interceptor.fulfilled, interceptor.rejected);
      });

      var responseInterceptorChain = [];
      this.interceptors.response.forEach(function pushResponseInterceptors(interceptor) {
        responseInterceptorChain.push(interceptor.fulfilled, interceptor.rejected);
      });

      var promise;

      if (!synchronousRequestInterceptors) {
        var chain = [dispatchRequest, undefined];

        Array.prototype.unshift.apply(chain, requestInterceptorChain);
        chain = chain.concat(responseInterceptorChain);

        promise = Promise.resolve(config);
        while (chain.length) {
          promise = promise.then(chain.shift(), chain.shift());
        }

        return promise;
      }


      var newConfig = config;
      while (requestInterceptorChain.length) {
        var onFulfilled = requestInterceptorChain.shift();
        var onRejected = requestInterceptorChain.shift();
        try {
          newConfig = onFulfilled(newConfig);
        } catch (error) {
          onRejected(error);
          break;
        }
      }

      try {
        promise = dispatchRequest(newConfig);
      } catch (error) {
        return Promise.reject(error);
      }

      while (responseInterceptorChain.length) {
        promise = promise.then(responseInterceptorChain.shift(), responseInterceptorChain.shift());
      }

      return promise;
    };

    Axios$1.prototype.getUri = function getUri(config) {
      config = mergeConfig$1(this.defaults, config);
      return buildURL(config.url, config.params, config.paramsSerializer).replace(/^\?/, '');
    };

    // Provide aliases for supported request methods
    utils$1.forEach(['delete', 'get', 'head', 'options'], function forEachMethodNoData(method) {
      /*eslint func-names:0*/
      Axios$1.prototype[method] = function(url, config) {
        return this.request(mergeConfig$1(config || {}, {
          method: method,
          url: url,
          data: (config || {}).data
        }));
      };
    });

    utils$1.forEach(['post', 'put', 'patch'], function forEachMethodWithData(method) {
      /*eslint func-names:0*/
      Axios$1.prototype[method] = function(url, data, config) {
        return this.request(mergeConfig$1(config || {}, {
          method: method,
          url: url,
          data: data
        }));
      };
    });

    var Axios_1 = Axios$1;

    var Cancel = Cancel_1;

    /**
     * A `CancelToken` is an object that can be used to request cancellation of an operation.
     *
     * @class
     * @param {Function} executor The executor function.
     */
    function CancelToken(executor) {
      if (typeof executor !== 'function') {
        throw new TypeError('executor must be a function.');
      }

      var resolvePromise;

      this.promise = new Promise(function promiseExecutor(resolve) {
        resolvePromise = resolve;
      });

      var token = this;

      // eslint-disable-next-line func-names
      this.promise.then(function(cancel) {
        if (!token._listeners) return;

        var i;
        var l = token._listeners.length;

        for (i = 0; i < l; i++) {
          token._listeners[i](cancel);
        }
        token._listeners = null;
      });

      // eslint-disable-next-line func-names
      this.promise.then = function(onfulfilled) {
        var _resolve;
        // eslint-disable-next-line func-names
        var promise = new Promise(function(resolve) {
          token.subscribe(resolve);
          _resolve = resolve;
        }).then(onfulfilled);

        promise.cancel = function reject() {
          token.unsubscribe(_resolve);
        };

        return promise;
      };

      executor(function cancel(message) {
        if (token.reason) {
          // Cancellation has already been requested
          return;
        }

        token.reason = new Cancel(message);
        resolvePromise(token.reason);
      });
    }

    /**
     * Throws a `Cancel` if cancellation has been requested.
     */
    CancelToken.prototype.throwIfRequested = function throwIfRequested() {
      if (this.reason) {
        throw this.reason;
      }
    };

    /**
     * Subscribe to the cancel signal
     */

    CancelToken.prototype.subscribe = function subscribe(listener) {
      if (this.reason) {
        listener(this.reason);
        return;
      }

      if (this._listeners) {
        this._listeners.push(listener);
      } else {
        this._listeners = [listener];
      }
    };

    /**
     * Unsubscribe from the cancel signal
     */

    CancelToken.prototype.unsubscribe = function unsubscribe(listener) {
      if (!this._listeners) {
        return;
      }
      var index = this._listeners.indexOf(listener);
      if (index !== -1) {
        this._listeners.splice(index, 1);
      }
    };

    /**
     * Returns an object that contains a new `CancelToken` and a function that, when called,
     * cancels the `CancelToken`.
     */
    CancelToken.source = function source() {
      var cancel;
      var token = new CancelToken(function executor(c) {
        cancel = c;
      });
      return {
        token: token,
        cancel: cancel
      };
    };

    var CancelToken_1 = CancelToken;

    /**
     * Syntactic sugar for invoking a function and expanding an array for arguments.
     *
     * Common use case would be to use `Function.prototype.apply`.
     *
     *  ```js
     *  function f(x, y, z) {}
     *  var args = [1, 2, 3];
     *  f.apply(null, args);
     *  ```
     *
     * With `spread` this example can be re-written.
     *
     *  ```js
     *  spread(function(x, y, z) {})([1, 2, 3]);
     *  ```
     *
     * @param {Function} callback
     * @returns {Function}
     */
    var spread = function spread(callback) {
      return function wrap(arr) {
        return callback.apply(null, arr);
      };
    };

    /**
     * Determines whether the payload is an error thrown by Axios
     *
     * @param {*} payload The value to test
     * @returns {boolean} True if the payload is an error thrown by Axios, otherwise false
     */
    var isAxiosError = function isAxiosError(payload) {
      return (typeof payload === 'object') && (payload.isAxiosError === true);
    };

    var utils = utils$d;
    var bind = bind$2;
    var Axios = Axios_1;
    var mergeConfig = mergeConfig$2;
    var defaults = defaults_1;

    /**
     * Create an instance of Axios
     *
     * @param {Object} defaultConfig The default config for the instance
     * @return {Axios} A new instance of Axios
     */
    function createInstance(defaultConfig) {
      var context = new Axios(defaultConfig);
      var instance = bind(Axios.prototype.request, context);

      // Copy axios.prototype to instance
      utils.extend(instance, Axios.prototype, context);

      // Copy context to instance
      utils.extend(instance, context);

      // Factory for creating new instances
      instance.create = function create(instanceConfig) {
        return createInstance(mergeConfig(defaultConfig, instanceConfig));
      };

      return instance;
    }

    // Create the default instance to be exported
    var axios$1 = createInstance(defaults);

    // Expose Axios class to allow class inheritance
    axios$1.Axios = Axios;

    // Expose Cancel & CancelToken
    axios$1.Cancel = Cancel_1;
    axios$1.CancelToken = CancelToken_1;
    axios$1.isCancel = isCancel$1;
    axios$1.VERSION = data.version;

    // Expose all/spread
    axios$1.all = function all(promises) {
      return Promise.all(promises);
    };
    axios$1.spread = spread;

    // Expose isAxiosError
    axios$1.isAxiosError = isAxiosError;

    axios$2.exports = axios$1;

    // Allow use of default import syntax in TypeScript
    axios$2.exports.default = axios$1;

    var axios = axios$2.exports;

    /**
     * Axios transport implementation
     */
    class AxiosTransport {
        constructor(url, storage, refresh = () => Promise.resolve()) {
            this._url = url;
            this._storage = storage;
            this._axios = null;
            this._refresh = refresh;
            this.url = url;
        }
        get url() {
            return this._url;
        }
        set url(value) {
            this._url = value;
            this._axios = axios.create({
                baseURL: value,
                withCredentials: true,
            });
        }
        get axios() {
            return this._axios;
        }
        get requests() {
            return {
                intercept: (onFulfilled, onRejected) => {
                    const id = this._axios.interceptors.request.use(onFulfilled, onRejected);
                    return {
                        eject: () => {
                            this._axios.interceptors.request.eject(id);
                        },
                    };
                },
            };
        }
        get responses() {
            return {
                intercept: (onFulfilled, onRejected) => {
                    const id = this._axios.interceptors.response.use(onFulfilled, onRejected);
                    return {
                        eject: () => {
                            this._axios.interceptors.response.eject(id);
                        },
                    };
                },
            };
        }
        request(method, path, data, options) {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j;
            return __awaiter(this, void 0, void 0, function* () {
                try {
                    options = options || {};
                    options.sendAuthorizationHeaders = (_a = options.sendAuthorizationHeaders) !== null && _a !== void 0 ? _a : true;
                    options.refreshTokenIfNeeded = (_b = options.refreshTokenIfNeeded) !== null && _b !== void 0 ? _b : true;
                    options.headers = (_c = options.headers) !== null && _c !== void 0 ? _c : {};
                    options.onUploadProgress = (_d = options.onUploadProgress) !== null && _d !== void 0 ? _d : undefined;
                    if (options.refreshTokenIfNeeded) {
                        yield this._refresh();
                    }
                    const config = {
                        method,
                        url: path,
                        data: data,
                        params: options.params,
                        headers: options.headers,
                        onUploadProgress: options.onUploadProgress,
                    };
                    const token = this._storage.auth_token;
                    const expiration = this._storage.auth_expires;
                    if (options.sendAuthorizationHeaders) {
                        if (token && ((expiration !== null && expiration > Date.now()) || expiration === null)) {
                            if (token.startsWith(`Bearer `)) {
                                config.headers.Authorization = token;
                            }
                            else {
                                config.headers.Authorization = `Bearer ${token}`;
                            }
                        }
                    }
                    const response = yield this.axios.request(config);
                    const responseData = response.data;
                    const content = {
                        raw: response.data,
                        status: response.status,
                        statusText: response.statusText,
                        headers: response.headers,
                        data: responseData.data,
                        meta: responseData.meta,
                        errors: responseData.errors,
                    };
                    if (responseData.errors) {
                        throw new TransportError(null, content);
                    }
                    return content;
                }
                catch (err) {
                    if (!err || err instanceof Error === false) {
                        throw err;
                    }
                    if (axios.isAxiosError(err)) {
                        const data = (_e = err.response) === null || _e === void 0 ? void 0 : _e.data;
                        throw new TransportError(err, {
                            raw: (_f = err.response) === null || _f === void 0 ? void 0 : _f.data,
                            status: (_g = err.response) === null || _g === void 0 ? void 0 : _g.status,
                            statusText: (_h = err.response) === null || _h === void 0 ? void 0 : _h.statusText,
                            headers: (_j = err.response) === null || _j === void 0 ? void 0 : _j.headers,
                            data: data === null || data === void 0 ? void 0 : data.data,
                            meta: data === null || data === void 0 ? void 0 : data.meta,
                            errors: data === null || data === void 0 ? void 0 : data.errors,
                        });
                    }
                    throw new TransportError(err);
                }
            });
        }
        get(path, options) {
            return __awaiter(this, void 0, void 0, function* () {
                return yield this.request('get', path, undefined, options);
            });
        }
        head(path, options) {
            return __awaiter(this, void 0, void 0, function* () {
                return yield this.request('head', path, undefined, options);
            });
        }
        options(path, options) {
            return __awaiter(this, void 0, void 0, function* () {
                return yield this.request('options', path, undefined, options);
            });
        }
        delete(path, data, options) {
            return __awaiter(this, void 0, void 0, function* () {
                return yield this.request('delete', path, data, options);
            });
        }
        put(path, data, options) {
            return __awaiter(this, void 0, void 0, function* () {
                return yield this.request('put', path, data, options);
            });
        }
        post(path, data, options) {
            return __awaiter(this, void 0, void 0, function* () {
                return yield this.request('post', path, data, options);
            });
        }
        patch(path, data, options) {
            return __awaiter(this, void 0, void 0, function* () {
                return yield this.request('patch', path, data, options);
            });
        }
    }

    class PasswordsHandler {
        constructor(transport) {
            this.transport = transport;
        }
        request(email, reset_url) {
            return __awaiter(this, void 0, void 0, function* () {
                yield this.transport.post('/auth/password/request', { email, reset_url });
            });
        }
        reset(token, password) {
            return __awaiter(this, void 0, void 0, function* () {
                yield this.transport.post('/auth/password/reset', { token, password });
            });
        }
    }

    class Debouncer {
        constructor(func) {
            this.func = func;
            this.debounced = [];
            this.debouncing = false;
        }
        debounce(...args) {
            return __awaiter(this, void 0, void 0, function* () {
                if (this.debouncing) {
                    return yield new Promise((resolve, reject) => {
                        this.debounced.push({
                            resolve: (value) => resolve(value),
                            reject: (error) => reject(error),
                        });
                    });
                }
                this.debouncing = true;
                return new Promise((resolve, reject) => {
                    this.func(...args)
                        .then((value) => {
                        const promises = [{ resolve, reject }, ...this.debounced];
                        this.debounced = [];
                        this.debouncing = false;
                        promises.forEach((promise) => promise.resolve(value));
                    })
                        .catch((error) => {
                        const promises = [{ resolve, reject }, ...this.debounced];
                        this.debounced = [];
                        this.debouncing = false;
                        promises.forEach((promise) => promise.reject(error));
                    });
                });
            });
        }
    }

    const DefaultExpirationTime = 30000;
    class Auth {
        constructor(transport, storage, options) {
            var _a, _b, _c, _d, _e;
            this.options = options || {};
            this.options.mode = (options === null || options === void 0 ? void 0 : options.mode) || (typeof window !== 'undefined' ? 'cookie' : 'json');
            this.options.refresh = (options === null || options === void 0 ? void 0 : options.refresh) || { auto: false, time: DefaultExpirationTime };
            this.options.refresh.auto = (_b = (_a = this.options.refresh) === null || _a === void 0 ? void 0 : _a.auto) !== null && _b !== void 0 ? _b : false;
            this.options.refresh.time = (_d = (_c = this.options.refresh) === null || _c === void 0 ? void 0 : _c.time) !== null && _d !== void 0 ? _d : DefaultExpirationTime;
            this.transport = transport;
            this.storage = storage;
            this.timer = false;
            this.refresher = new Debouncer(this.refreshToken.bind(this));
            try {
                this.updateRefresh((_e = this.options) === null || _e === void 0 ? void 0 : _e.refresh);
            }
            catch (_f) {
                // Ignore error
            }
        }
        get token() {
            return this.storage.auth_token;
        }
        get password() {
            return (this.passwords = this.passwords || new PasswordsHandler(this.transport));
        }
        get expiring() {
            var _a, _b;
            const expiration = this.storage.auth_expires;
            if (expiration === null) {
                return false;
            }
            const expiringAfter = expiration - ((_b = (_a = this.options.refresh) === null || _a === void 0 ? void 0 : _a.time) !== null && _b !== void 0 ? _b : 0);
            return expiringAfter <= Date.now();
        }
        refreshToken(force = false) {
            var _a;
            return __awaiter(this, void 0, void 0, function* () {
                if (!force && !this.expiring) {
                    return false;
                }
                const response = yield this.transport.post('/auth/refresh', {
                    refresh_token: this.options.mode === 'json' ? this.storage.auth_refresh_token : undefined,
                }, {
                    refreshTokenIfNeeded: false,
                });
                this.updateStorage(response.data);
                this.updateRefresh();
                return {
                    access_token: response.data.access_token,
                    refresh_token: (_a = response.data) === null || _a === void 0 ? void 0 : _a.refresh_token,
                    expires: response.data.expires,
                };
            });
        }
        updateStorage(result) {
            var _a;
            this.storage.auth_token = result.access_token;
            this.storage.auth_refresh_token = (_a = result.refresh_token) !== null && _a !== void 0 ? _a : null;
            if (result.expires) {
                this.storage.auth_expires = Date.now() + result.expires;
            }
            else {
                this.storage.auth_expires = null;
            }
        }
        updateRefresh(options) {
            var _a, _b;
            const expiration = this.storage.auth_expires;
            if (expiration === null) {
                clearTimeout(this.timer);
                return; // Don't auto refresh if there's no expiration time (token auth)
            }
            if (options) {
                this.options.refresh.auto = (_a = options.auto) !== null && _a !== void 0 ? _a : this.options.refresh.auto;
                this.options.refresh.time = (_b = options.time) !== null && _b !== void 0 ? _b : this.options.refresh.time;
            }
            clearTimeout(this.timer);
            let remaining = expiration - this.options.refresh.time - Date.now();
            if (remaining < 0) {
                // It's already expired, try a refresh
                if (expiration < Date.now()) {
                    return; // Don't set auto refresh
                }
                else {
                    remaining = 0;
                }
            }
            if (this.options.refresh.auto) {
                this.timer = setTimeout(() => {
                    this.refresh()
                        .then(() => {
                        // Do nothing
                    })
                        .catch(() => {
                        // Do nothing
                    });
                }, remaining);
            }
        }
        refresh(force = false) {
            return __awaiter(this, void 0, void 0, function* () {
                return yield this.refresher.debounce(force);
            });
        }
        login(credentials, options) {
            var _a;
            return __awaiter(this, void 0, void 0, function* () {
                options = options || {};
                const response = yield this.transport.post('/auth/login', Object.assign({ mode: this.options.mode }, credentials), {
                    refreshTokenIfNeeded: false,
                    sendAuthorizationHeaders: false,
                });
                this.updateStorage(response.data);
                this.updateRefresh(options.refresh);
                return {
                    access_token: response.data.access_token,
                    refresh_token: (_a = response.data) === null || _a === void 0 ? void 0 : _a.refresh_token,
                    expires: response.data.expires,
                };
            });
        }
        static(token) {
            return __awaiter(this, void 0, void 0, function* () {
                yield this.transport.get('/users/me', {
                    params: {
                        access_token: token,
                    },
                });
                this.storage.auth_token = token;
                this.storage.auth_expires = null;
                this.storage.auth_refresh_token = null;
                return true;
            });
        }
        logout() {
            return __awaiter(this, void 0, void 0, function* () {
                let refresh_token;
                if (this.options.mode === 'json') {
                    refresh_token = this.storage.auth_refresh_token || undefined;
                }
                yield this.transport.post('/auth/logout', {
                    refresh_token,
                }, {
                    refreshTokenIfNeeded: false,
                });
                this.storage.auth_token = null;
                this.storage.auth_expires = null;
                this.storage.auth_refresh_token = null;
                clearTimeout(this.timer);
            });
        }
    }

    class GraphQLHandler {
        constructor(transport) {
            this.transport = transport;
        }
        request(base, query, variables) {
            return __awaiter(this, void 0, void 0, function* () {
                return yield this.transport.post(base, {
                    query,
                    variables: typeof variables === 'undefined' ? {} : variables,
                });
            });
        }
        // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
        items(query, variables) {
            return __awaiter(this, void 0, void 0, function* () {
                return yield this.request('/graphql', query, variables);
            });
        }
        // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
        system(query, variables) {
            return __awaiter(this, void 0, void 0, function* () {
                return yield this.request('/graphql/system', query, variables);
            });
        }
    }

    class Directus {
        constructor(url, options) {
            this._storage = (options === null || options === void 0 ? void 0 : options.storage) || (typeof window !== 'undefined' ? new LocalStorage() : new MemoryStorage());
            this._transport =
                (options === null || options === void 0 ? void 0 : options.transport) ||
                    new AxiosTransport(url, this._storage, () => __awaiter(this, void 0, void 0, function* () {
                        yield this._auth.refresh();
                    }));
            this._auth = (options === null || options === void 0 ? void 0 : options.auth) || new Auth(this._transport, this._storage);
            this._items = {};
            this._singletons = {};
        }
        get auth() {
            return this._auth;
        }
        get storage() {
            return this._storage;
        }
        get transport() {
            return this._transport;
        }
        get activity() {
            return this._activity || (this._activity = new ActivityHandler(this.transport));
        }
        get collections() {
            return (this._collections ||
                (this._collections = new CollectionsHandler(this.transport)));
        }
        get fields() {
            return this._fields || (this._fields = new FieldsHandler(this.transport));
        }
        get files() {
            return this._files || (this._files = new FilesHandler(this.transport));
        }
        get folders() {
            return this._folders || (this._folders = new FoldersHandler(this.transport));
        }
        get permissions() {
            return (this._permissions ||
                (this._permissions = new PermissionsHandler(this.transport)));
        }
        get presets() {
            return this._presets || (this._presets = new PresetsHandler(this.transport));
        }
        get relations() {
            return this._relations || (this._relations = new RelationsHandler(this.transport));
        }
        get revisions() {
            return this._revisions || (this._revisions = new RevisionsHandler(this.transport));
        }
        get roles() {
            return this._roles || (this._roles = new RolesHandler(this.transport));
        }
        get users() {
            return this._users || (this._users = new UsersHandler(this.transport));
        }
        get settings() {
            return this._settings || (this._settings = new SettingsHandler(this.transport));
        }
        get server() {
            return this._server || (this._server = new ServerHandler(this.transport));
        }
        get utils() {
            return this._utils || (this._utils = new UtilsHandler(this.transport));
        }
        get graphql() {
            return this._graphql || (this._graphql = new GraphQLHandler(this.transport));
        }
        singleton(collection) {
            return (this._singletons[collection] ||
                (this._singletons[collection] = new SingletonHandler(collection, this.transport)));
        }
        items(collection) {
            return this._items[collection] || (this._items[collection] = new ItemsHandler(collection, this.transport));
        }
    }

    exports.ActivityHandler = ActivityHandler;
    exports.Auth = Auth;
    exports.AxiosTransport = AxiosTransport;
    exports.BaseStorage = BaseStorage;
    exports.CollectionsHandler = CollectionsHandler;
    exports.CommentsHandler = CommentsHandler;
    exports.Directus = Directus;
    exports.FieldsHandler = FieldsHandler;
    exports.FilesHandler = FilesHandler;
    exports.FoldersHandler = FoldersHandler;
    exports.ItemsHandler = ItemsHandler;
    exports.LocalStorage = LocalStorage;
    exports.MemoryStorage = MemoryStorage;
    exports.PermissionsHandler = PermissionsHandler;
    exports.PresetsHandler = PresetsHandler;
    exports.RelationsHandler = RelationsHandler;
    exports.RevisionsHandler = RevisionsHandler;
    exports.RolesHandler = RolesHandler;
    exports.ServerHandler = ServerHandler;
    exports.SettingsHandler = SettingsHandler;
    exports.TransportError = TransportError;
    exports.UsersHandler = UsersHandler;
    exports.UtilsHandler = UtilsHandler;

    Object.defineProperty(exports, '__esModule', { value: true });

    return exports;

})({});
